driver/convert.c

```diff
@@ -1472,37 +1475,37 @@ static SQLRETURN wstr_to_cstr(esodbc_rec_st *arec, esodbc_rec_st *irec,
 		 * conversion function. */
 		in_bytes = (int)buff_octet_size(in_bytes, sizeof(SQLCHAR), arec, irec,
 				&state);
-		/* trim the original string until it fits in output buffer, with given
-		 * length limitation */
-		for (c = (int)xstr.w.cnt + 1; 0 < c; c --) {
-			out_bytes = U16WC_TO_MBU8(xstr.w.str, c, charp, in_bytes);
-			/* if user gives 0 as buffer size, out_bytes will also be 0 */
-			if (out_bytes <= 0) {
-				if (WAPI_ERR_EBUFF()) {
-					continue;
-				}
-				ERRNH(stmt, "failed to convert wchar_t* to char* for string `"
-					LWPDL "`.", c, xstr.w.str);
-				RET_HDIAGS(stmt, SQL_STATE_22018);
-			} else {
-				/* conversion succeeded */
-				break;
+		if (in_bytes) {
+			/* trim the original string until it fits in output buffer, with
+			 * given length limitation */
+			for (c = (int)xstr.w.cnt + 1; 0 < c; c --) {
+				out_bytes = U16WC_TO_MBU8(xstr.w.str, c, charp, in_bytes);
+				if (0 < out_bytes) {
+					break; /* conversion succeeded */
+				} // else: out_bytes <= 0
+				if (! WAPI_ERR_EBUFF()) {
+					ERRNH(stmt, "failed to convert wchar_t* to char* for "
+						"string `" LWPDL "`.", c, xstr.w.str);
+					RET_HDIAGS(stmt, SQL_STATE_22018);
+				} // else: buffer too small for full string: trimm further
 			}
-		}
 
-		assert(0 < out_bytes);
-		if (charp[out_bytes - 1]) {
-			/* ran out of buffer => not 0-terminated and truncated already */
-			charp[out_bytes - 1] = 0;
-			state = SQL_STATE_01004; /* indicate truncation */
-			c --; /* last char was overwritten with 0 -> dec xfed count */
-		}
-
-		/* only update offset if data is copied out */
-		gd_offset_update(stmt, xstr.w.cnt, c);
+			assert(0 < out_bytes);
+			if (charp[out_bytes - 1] != '\0') {
+				/* ran out of buffer => not 0-term'd and truncated already */
+				charp[out_bytes - 1] = 0;
```

Comment:
nit: in the test you used `'\0'` but on this line you're using `0`.  It shouldn't make any difference but might be nice to consistently use one or the other.

Reply:
Thanks, fixed them.